crypto/bio/bss_dgram.c

```diff
@@ -808,6 +992,627 @@ static int dgram_puts(BIO *bp, const char *str)
+        translate_msg(&mh, &iov, control, &BIO_MSG_N(msg, i));
+
+        /* If local address was requested, it must have been enabled */
+        if (BIO_MSG_N(msg, i).local != NULL && !have_local_enabled)
+            /*
+             * If we have done at least one message, we must return the
+             * count; if we haven't done any, we can give an error code
+             */
+            return i > 0 ? (ossl_ssize_t)i : -1;
+
+        ret = recvmsg(b->num, &mh, sysflags);
+        if (ret <= 0)
+            break;
+
+        BIO_MSG_N(msg, i).data_len = ret;
+        BIO_MSG_N(msg, i).flags    = 0;
+        if (BIO_MSG_N(msg, i).local != NULL)
+            if (extract_local(b, &mh, BIO_MSG_N(msg, i).local) < 1)
+                return i > 0 ? (ossl_ssize_t)i : -1;
+    }
+
+    return i > 0 ? (ossl_ssize_t)i : PACK_ERRNO(get_last_socket_error());
+
+#elif M_METHOD == M_METHOD_RECVFROM || M_METHOD == M_METHOD_WSARECVMSG
+# if M_METHOD == M_METHOD_WSARECVMSG
+    if (bio_WSARecvMsg != NULL) {
+        /* WSARecvMsg-based implementation for Windows. */
+        translate_msg_win(&wmsg, &wbuf, control, msg);
+
+        /* If local address was requested, it must have been enabled */
+        if (msg[0].local != NULL && !have_local_enabled)
+            return -1;
+
+        ret = WSARecvMsg((SOCKET)b->num, &wmsg, &num_bytes_received, NULL, NULL);
+        if (ret < 0)
+            return PACK_ERRNO(get_last_socket_error());
+
+        msg[0].data_len = num_bytes_received;
+        msg[0].flags    = 0;
+        if (msg[0].local != NULL)
+            if (extract_local(b, &wmsg, msg[0].local) < 1)
+                /*
+                 * On Windows, loopback is not a "proper" interface and it works
+                 * differently; packets are essentially short-circuited and
+                 * don't go through all of the normal processing. A consequence
+                 * of this is that packets sent from the local machine to the
+                 * local machine _will not have IP_PKTINFO_ even if the
+                 * IP_PKTINFO socket option is enabled. WSARecvMsg just sets
+                 * Control.len to 0 on returning.
+                 *
+                 * This applies regardless of whether the loopback address,
+                 * 127.0.0.1 is used, or a local interface address (e.g.
+                 * 192.168.1.1); in both cases IP_PKTINFO will not be present.
+                 *
+                 * We report this condition by setting the local BIO_ADDR's
+                 * family to 0.
+                 */
+                BIO_ADDR_clear(msg[0].local);
+
+        return 1;
+    }
+# endif
+
+    /*
+     * Fallback to recvfrom and receive a single message.
+     */
+    if (msg[0].local != NULL)
+        /*
+         * We cannot determine the local address if using recvfrom
+         * so fail in this case
+         */
+        return -1;
+
+    slen = sizeof(*msg[0].peer);
+    ret = recvfrom(b->num, msg[0].data,
```

Comment:
Why do we not simulate `recvmmsg` here by calling this in a loop like we do for `recvmsg`?